emote/algorithms/amp.py

```diff
@@ -0,0 +1,182 @@
+from emote.callbacks.loss import LossCallback
+
+
+def gradient_loss_function(model_output: Tensor, model_input: Tensor) -> Tensor:
+    """
+    Given inputs and outputs of an nn.Module, computes the sum of
+    squared derivatives of outputs to the inputs
+        Arguments:
+            model_output (Tensor): the output of the nn.Module
+            model_input (Tensor): the input to the nn.Module
+        Returns:
+            loss (Tensor): the sum of squared derivatives
+    """
+    # grad can be implicitly created only for scalar outputs
+    predictions = torch.split(model_output, 1, dim=0)
+    inputs_grad = torch.autograd.grad(
+        predictions, model_input, create_graph=True, retain_graph=True
+    )
+    inputs_grad = torch.cat(inputs_grad, dim=1)
+    inputs_grad_norm = torch.square(inputs_grad)
+    inputs_grad_norm = torch.sum(inputs_grad_norm, dim=1)
+    return torch.mean(inputs_grad_norm)
+
+
+class DiscriminatorLoss(LossCallback):
+    """
+    This loss is used to train a discriminator for
+    adversarial training.
+    """
+
+    def __init__(
+        self,
+        discriminator: nn.Module,
+        state_map_fn: Callable[[Tensor], Tensor],
+        grad_loss_weight: float,
+        optimizer: torch.optim.Optimizer,
+        lr_schedule: torch.optim.lr_scheduler._LRScheduler,
+        max_grad_norm: float,
+        input_key: str = "features",
+        name: str = "Discriminator",
+    ):
+        super().__init__(
+            lr_schedule=lr_schedule,
+            name=name,
+            network=discriminator,
+            optimizer=optimizer,
+            max_grad_norm=max_grad_norm,
+            data_group=None,
+        )
+        self._discriminator = discriminator
+        self._state_map_function = state_map_fn
+        self._grad_loss_weight = grad_loss_weight
+        self._obs_key = input_key
+
+    def loss(self, imitation_batch: dict, policy_batch: dict) -> Tensor:
+        """
+        Computing the loss
+            Arguments:
+                imitation_batch (dict): a batch of data from the reference animation.
+                    the discriminator is trained to classify data from this batch as
+                    positive samples
+                policy_batch (dict): a batch of data from the RL buffer. the discriminator
+                    is trained to classify data from this batch as negative samples.
+            Returns:
+                loss (Tensor): the loss tensor
+        """
+        imitation_batch_size = imitation_batch["batch_size"]
+        policy_data_batch_size = policy_batch["batch_size"]
+
+        pos_obs: Tensor = imitation_batch["observation"][self._obs_key]
+        pos_next_obs: Tensor = imitation_batch["next_observation"][self._obs_key]
+        neg_obs: Tensor = policy_batch["observation"][self._obs_key]
+        neg_next_obs: Tensor = policy_batch["next_observation"][self._obs_key]
+
+        (pos_obs,) = (self._state_map_function(pos_obs),)
```

Comment:
why the tuple wrap?

Reply:
Have no idea how this happened! It should be some automatic code formatting. Thanks for the catch :)